include/revng/Model/Types.h

```diff
@@ -0,0 +1,903 @@
+
+  static flatc::StructField fromKey(const std::size_t &Offset) {
+    return flatc::StructField(Offset);
+  }
+};
+
+/// \brief A struct type in flatc. Structs are actually typedefs of unnamed
+/// structs in C.
+struct flatc::StructType : public flatc::Type {
+  static constexpr const char *Tag = "!flatc::StructType";
+
+  // StructType is kinda weird, because we have a set of conflicting
+  // requirements:
+  // - first of all, AggregateFields contains a TupleTreeReference, so we
+  // cannot
+  //   hold fields in a std::vector
+  //   - this is due to the fact that TupleTree still does not handle
+  //   automatic initialization of TupleTreeReferences that are inside a
+  //   std::vector
+  //   - so we need to hold the fields in a supported container. likely a
+  //   SortedVector. but we have to decide the key that will be used to order
+  //   them
+  // - we want to be able to index fields by name
+  //   - this is a hint that the Name should be the key for fields
+  //   - this is also consistent with the fact that in C it's not possible to
+  //   have 2 different fields of a struct with the same name
+  //   - but accessing fields by name is rarely necessary in practice
+  //   - also, renaming a field changes its order
+  // - most times we want to iterate on fields we actually want to do it in
+  //   positional order, meaning that fields that come first in the actual
+  //   layout of the struct would need to be iterated over sooner
+  //   - this suggests that we want the key for the fields to be an integer
+  //   representing their position in the struct
+  //   - but if we remove a field from the struct after its creation, we end
+  //   up with a missing index, which is a thing that we'd like to avoid
+  //   - this kind of iteration is important for actually computing pointer
+  //   arithmetic and member access expressions when generating C code, and
+  //   it's likely the most common operation
+
+  SortedVector<StructField> Fields;
+  std::size_t Size;
+
+  StructType(llvm::StringRef NameRef) :
+    Type(NameRef, TypeKind::Struct), Fields() {}
+
+  StructType() : StructType("") {}
+  ~StructType() = default;
+
+  StructType(const StructType &) = default;
+  StructType &operator=(const StructType &) = default;
+
+  StructType(StructType &&) = default;
+  StructType &operator=(StructType &&) = default;
+
+  static bool classof(const Type *T) {
+    return T->getKind() == TypeKind::Struct;
+  }
+
+  RecursiveCoroutine<std::size_t> size() const;
+
+private:
+  // Kind must be TypeKind::Struct, Fields must not be empty, all Fields must
+  // verify and Fields must not have duplicate names.
+  bool verifyDerived() const debug_function;
+
+  friend struct flatc::Type;
+};
+INTROSPECTION_NS(flatc, StructType, UUID, Name, Kind, Fields, Size);
+
+/// \brief A field of a union type in flatc, with position, qualified type, and
+/// name
+struct flatc::UnionField : flatc::AggregateField {
+  std::size_t Position;
```

Comment:
Why is `Position` the key of `union`?

Reply:
Mainly because we want to be able to rename fields, so the name cannot be the key. It could be some other kind of ID, but in this way it's related to the order of printing. Fields with lower position are printed first. Which makes easier to write a valid union in yaml by hand.

Thinking about this, the position is not necessarily the best solution anyway. If we remove, say, the third field of a union, what are we going to do? Rescale others is something we don't want to do, cause we should change their keys and it would mess up the diff.

So at this point I'm thinking we should relax the `verify()` on `UnionType` so that it allows fields with non-consecutive positions. In this way we don't need to rescale on removal, and we don't break diffs.

Otherwise we could replace the `Position` with some other kind of ID, like a UUID, but I'm not particularly fond of it.

I have no strong feelings here. The point is all the solutions we can devise will be somehow counter-intuitive, because on one hand we'd like to have the name as a key, but that's a no-go if we want to enable renaming, which is something we'll need to do for sure. So it's a matter of choosing a counter-intuitive solution that is less wrong than the others.

Reply:
I think that we should deal with the fact that, wherever we want to put a name as key, we have to use a UUID.

Reply:
Ok, will do.

Reply:
Done